pkg/token/serviceprincipaltoken_test.go

```diff
@@ -0,0 +1,232 @@
+package token
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"gopkg.in/dnaeon/go-vcr.v3/cassette"
+	"gopkg.in/dnaeon/go-vcr.v3/recorder"
+)
+
+const (
+	clientID          = "AZURE_CLIENT_ID"
+	clientSecret      = "AAD_SERVICE_PRINCIPAL_CLIENT_SECRET"
+	clientCert        = "AZURE_CLIENT_CER"
+	clientCertPass    = "AZURE_CLIENT_CERTIFICATE_PASSWORD"
+	resourceID        = "AZURE_RESOURCE_ID"
+	tenantID          = "AZURE_TENANT_ID"
+	vcrMode           = "VCR_MODE"
+	vcrModeRecordOnly = "RecordOnly"
+	badSecret         = "Bad_Secret"
+	redactionToken    = "[REDACTED]"
+)
+
+func TestMissingLoginMethods(t *testing.T) {
+	p := &servicePrincipalToken{}
+	expectedErr := "service principal token requires either client secret or certificate"
+
+	_, err := p.Token()
+	if !ErrorContains(err, expectedErr) {
+		t.Errorf("expected error %v, but got %v", expectedErr, err)
+	}
+}
+
+func TestMissingCertFile(t *testing.T) {
+	p := &servicePrincipalToken{
+		clientCert: "testdata/noCertHere.pfx",
+	}
+	expectedErr := "failed to read the certificate file"
+
+	_, err := p.Token()
+	if !ErrorContains(err, expectedErr) {
+		t.Errorf("expected error %v, but got %v", expectedErr, err)
+	}
+}
+
+func TestBadCertPassword(t *testing.T) {
+	p := &servicePrincipalToken{
+		clientCert:         "testdata/testCert.pfx",
+		clientCertPassword: badSecret,
+	}
+	expectedErr := "failed to decode pkcs12 certificate while creating spt: pkcs12: decryption password incorrect"
+
+	_, err := p.Token()
+	if !ErrorContains(err, expectedErr) {
+		t.Errorf("expected error %v, but got %v", expectedErr, err)
+	}
+}
+
+// getVCRHttpClient setup Go-vcr
+func getVCRHttpClient(path string) (*recorder.Recorder, *http.Client) {
```

Comment:
can you separate these go-vcr utility functions into its own file?

Reply:
Most of the go-vcr functions are highly customized for service principal tests. I'm hesitant to put them in their own file without being more general.

Reply:
Done